clusterloader2/pkg/imagepreload/imagepreload.go

```diff
@@ -0,0 +1,166 @@
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/klog"
+	"k8s.io/perf-tests/clusterloader2/pkg/config"
+	"k8s.io/perf-tests/clusterloader2/pkg/flags"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework"
+	"k8s.io/perf-tests/clusterloader2/pkg/framework/client"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util"
+	"k8s.io/perf-tests/clusterloader2/pkg/measurement/util/informer"
+)
+
+const (
+	informerTimeout = time.Minute
+	manifest        = "$GOPATH/src/k8s.io/perf-tests/clusterloader2/pkg/imagepreload/manifests/daemonset.yaml"
+	namespace       = "preload"
+	pollingInterval = 5 * time.Second
+	pollingTimeout  = 5 * time.Minute
+)
+
+var images []string
+
+func init() {
+	flags.StringSliceEnvVar(&images, "node-preload-images", "NODE_PRELOAD_IMAGES", []string{}, "List of images to preload on each node in the test cluster before executing tests")
+}
+
+type controller struct {
+	config          *config.ClusterLoaderConfig
+	framework       *framework.Framework
+	templateMapping map[string]interface{}
+	images          []string
+
+	lock sync.Mutex
+}
+
+// Setup ensures every node in cluster preloads given list of images before starting tests.
+// It does it by creating a Daemonset that call "docker pull" and awaits for Node object to be updated.
+// As a side-effect of the image preloading, size of Node objects is increased.
+//
+// Preloading is skipped in kubemark or if no images have been specified.
+func Setup(conf *config.ClusterLoaderConfig, f *framework.Framework) error {
+	mapping, err := config.GetMapping(conf)
+	if err != nil {
+		return err
+	}
+
+	ctl := &controller{
+		config:          conf,
+		framework:       f,
+		templateMapping: mapping,
+		images:          images,
+	}
+	return ctl.PreloadImages()
+}
+
+func (c *controller) PreloadImages() error {
+	if len(images) == 0 {
+		klog.Warning("No images specified. Skipping image preloading")
+		return nil
+	}
+	if c.config.ClusterConfig.Provider == "kubemark" {
+		klog.Warning("Image preloading is disabled in kubemark")
+		return nil
+	}
+
+	clusterSize := c.config.ClusterConfig.Nodes
+	kclient := c.framework.GetClientSets().GetClient()
+
+	doneNodes := make(map[string]bool)
+	stopCh := make(chan struct{})
+	defer close(stopCh)
+
+	ni := informer.NewInformer(
+		kclient,
+		"nodes",
+		util.NewObjectSelector(),
+		func(old, new interface{}) { c.checkNode(new, doneNodes) })
+	if err := informer.StartAndSync(ni, stopCh, informerTimeout); err != nil {
+		return err
+	}
+
+	klog.Infof("Creating namespace %s...", namespace)
+	if err := client.CreateNamespace(kclient, namespace); err != nil {
+		return err
+	}
+
+	klog.Info("Creating daemonset to preload images...")
+	c.templateMapping["Images"] = c.images
+	if err := c.framework.ApplyTemplatedManifests(manifest, c.templateMapping); err != nil {
+		return err
+	}
+
+	klog.Info("Waiting for Node objects to be updated...")
+	if err := wait.Poll(pollingInterval, pollingTimeout, func() (bool, error) {
```

Comment:
Are we sure pollingtimeout = 5m will be enough?

Reply:
We use this value in other places. Hard to tell to be honest, but I'd suspect this to be fine. We can always bump it if needed.

Reply:
What places? Here we wait until N pods will download M images - I don't think we have precedent for that anywhere.
At least add a TODO to validate this value.

Reply:
For that specifically defined task no - we don't have. But in other places (e.g. Prometheus). I'll add a note on that.